keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,285 @@
+1. A user creates a Bucket in the app namespace, specifying the BucketClass.
+1. The COSI controller detects the Bucket and creates a BucketContent object in the `cosi-system` namespace.
+1. The sidecar detects the BucketContent object and calls the GrantAccess() rpc to the driver, returing a set of credentials for accessing the bucket.
+1. The sidecar writes the credentials and endpoint information to a Secret in the `cosi-system` namespace, with an owner reference to the BucketContent.
+1. The COSI controller generates a Secret in the Bucket's namespace containing the parent Secret's data, with an owner reference to the Bucket.
+1. The workflow ingests the Secret to begin operation.
+
+#### Revoke Access
+
+1. The user deletes the Bucket, which blocks until revoke operations complete.
+2. The COSI controller detects the update and deletes the BucketContent, which also blocks until backend revoke operations complete.
+3. The sidecar detects the BucketContent update and invokes the RevokeAccess() rpc method.
+4. The driver terminates the access for the associated credentials.
+5. The sidecar removes the finalizer from the BucketContent, allowing it and the parent Secret to be garbage collected.
+6. The COSI controller removes the finalizer from the Bucket,  allowing it and the child Secret to be garbage collected.
+
+### Static Buckets
+
+> Note: No driver is present to manage provisioning.  The COSI controller only automates Bucket/BucketContent binding operations.
+
+#### Grant Access
+
+1. A BucketClass is defined specifically referencing an object store bucket and a Secret in the `cosi-system` namespace containing access credentials.
+1. A user creates a Bucket in the app namespace, specifying the BucketClass.
+1. The COSI controller detects the object store bucket and Secret in the BucketClass and creates a BucketContent in the `cosi-system` namespace.
+1. The COSI controller generates a Secret in the Bucket's namespace containing the parent Secret's data, with an owner reference to the Bucket.
+1. The workflow ingests the Secret to begin operation.
+
+#### Revoke Access
+
+1. The user deletes the Bucket, which blocks until revoke operations complete.
+1. The COSI controller deletes the Bucket's bound BucketContent object.
+1. The COSI controller removes the finalizer from the Bucket, allowing it and the child Secret to be garbage collected.
+
+##  Custom Resource Definitions
+
+#### Bucket
+
+A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the Bucket is "bound" to the corresponding BucketContent. The is used to prevent further binds to the BucketContent.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  bucketName: [3]
+  generateBucketName: [4]
+  bucketClassName: [5]
+  secretName: [6]
+  protocol: [7]
```

Comment:
since it's an array should it be named `protocols`?

Reply:
Is it reasonable to think that an app would be written around more than one client sdk?